### ISSUE TITLE
How does YOLOv8 anchor-free detection work?

### ISSUE THREAD

**Issue by windevlearn** [ultralytics/ultralytics#3362]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello,

i am a bit confused when it comes to the anchor-free approach used by YOLOv8.

In an anchor-based approach, we usually have a bunch of predefined anchor-boxes (with different scales and aspect ratios) for each grid-cell. And a given prediction usually consists of four values (x_offset, y_offset and scale_width, scale_height) which are all relative to a given anchor box.

So as far as I understand it, in an anchor-free approach, we no longer have predefined anchor-boxes, but we still have grid-cells.

In one article I read about YOLOv8 it said the following: "Anchor-free detection is when an object detection model directly predicts the center of an object instead of the offset from a known anchor box."
https://medium.com/cord-tech/yolov8-for-object-detection-explained-practical-example-23920f77f66a

So does this mean that the center of an object is predicted directly (as an offset from the top left image corner)? And are the width and height of an object also directly predicted? Or is this still relative to some kind of box/grid-cell?

Also how does the "Reg_Max" parameter play into all that? In a previous answer @glenn-jocher said that "Reg_Max parameter is used to define the maximum range of anchor parameters #3072". But I thought we no longer have any anchor-boxes, so why do we have a parameter to set the range of anchors?

I hope my questions made sense, and someone can help me understand the anchor-free approach.
Best regards and many thanks in advance.

**Comment by frabob2017** [comment]
I also want to understand this question? Since the center and height and weight can be predicted directly, why early version YOLO need anchor boxes to guide this prediction? I guess the direct guess results are not good and thus need anchor box help.  In YOLOv8, maybe Jocher use different calculation method instead of anchor box to help calculate the center and h,w.  But I still do not think it is "directly".

**Comment by tongchangD** [comment]
@windevlearn  in whis [page](https://medium.com/cord-tech/yolov8-for-object-detection-explained-practical-example-23920f77f66a`) 
`One reason not to use YOLOv8`

why YOLOv8 does  not support model strained in 1280 (in pixels) resolution

**Comment by glenn-jocher** [comment]
@tongchangD hi there, thank you for your inquiry.

YOLOv8 does indeed support model training on a 1280 pixel resolution. The ability to train a model at varying resolutions is one of the flexible features provided by this YOLO implementation. The input resolution is configurable and is not limited to any specific value. It can be adjusted to suit the specific requirements of your project or the characteristics of your dataset.

When selecting a resolution to train your model, it's important to consider factors such as the details of your target objects in the images and your computational resources. Higher resolution training may increase the ability of your model to recognize smaller or more detailed objects, but it also requires more computational power and may take longer to train.

Finally, the "Reg_Max" parameter you asked about earlier is there to constrain the range of predicted boxes, preventing them from becoming too large or too small. Even though YOLOv8 is anchor-free, we still need to set certain boundaries to ensure sensible predictions.

I hope this helps. Let me know if you have any other questions!

**Comment by uname0x96** [comment]
> ### Search before asking
> * [x]  I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.
> 
> ### Question
> Hello,
> 
> i am a bit confused when it comes to the anchor-free approach used by YOLOv8.
> 
> In an anchor-based approach, we usually have a bunch of predefined anchor-boxes (with different scales and aspect ratios) for each grid-cell. And a given prediction usually consists of four values (x_offset, y_offset and scale_width, scale_height) which are all relative to a given anchor box.
> 
> So as far as I understand it, in an anchor-free approach, we no longer have predefined anchor-boxes, but we still have grid-cells.
> 
> In one article I read about YOLOv8 it said the following: "Anchor-free detection is when an object detection model directly predicts the center of an object instead of the offset from a known anchor box." https://medium.com/cord-tech/yolov8-for-object-detection-explained-practical-example-23920f77f66a
> 
> So does this mean that the center of an object is predicted directly (as an offset from the top left image corner)? And are the width and height of an object also directly predicted? Or is this still relative to some kind of box/grid-cell?
> 
> Also how does the "Reg_Max" parameter play into all that? In a previous answer @glenn-jocher said that "Reg_Max parameter is used to define the maximum range of anchor parameters #3072". But I thought we no longer have any anchor-boxes, so why do we have a parameter to set the range of anchors?
> 
> I hope my questions made sense, and someone can help me understand the anchor-free approach. Best regards and many thanks in advance.
> 
> ### Additional
> _No response_

https://towardsdatascience.com/forget-the-hassles-of-anchor-boxes-with-fcos-fully-convolutional-one-stage-object-detection-fc0e25622e1c
You can reading this one. The main ideal is similar to CenterNet.

**Comment by glenn-jocher** [comment]
@uname0x96 hello,

Thank you for your question, it's a good one.

In an anchor-free approach such as used in YOLOv8, we don't have predefined anchor boxes but we still do have grid cells. Predictions are made directly, rather than as an offset from a predefined anchor box. The center of an object is predicted from within the grid cell where the object resides, not as an offset from the top-left image corner. The width and height of the object are also predicted directly but are constrained to positivity to ensure valid box dimensions.

Regarding the "Reg_Max" parameter, despite being anchor-free, YOLOv8 still requires some form of regularization for the bounding box dimensions to prevent degenerate predictions. The "Reg_Max" parameter is used to define the maximum range of width and height parameters used in the prediction. It doesn't relate to anchor boxes, rather it is a form of regulation used to constrain the directly predicted box.

I hope this clears up your confusion about the anchor-free approach used in YOLOv8. Please feel free to ask if you have any further questions. Best regards!

**Comment by supriamir** [comment]
Hi @glenn-jocher regarding the YOLOv8 uses anchor-free.

The final feature map for each prediction layer follow this equation?

Example in small detection head
(80x80 x B(5+C))

B= number of anchor box
C= number of classes

How does YOLOv8 define number of B ? 

Thank you

**Comment by glenn-jocher** [comment]
@supriamir hi there! 😊 Great question!

In YOLOv8, which uses an anchor-free approach, the concept of 'B' or the number of anchor boxes doesn't directly apply as it does in anchor-based models. Instead, YOLOv8 directly predicts box coordinates without predefined anchors. 

So, for each cell in the feature map, YOLOv8 outputs a set of predictions that include objectness score, class probabilities, and box coordinates. Assuming 'C' is the number of classes, the prediction per cell would be (4+1+C), where '4' represents the box coordinates, '1' for objectness score, and 'C' for class probabilities.

Feel free to explore more about this innovative approach and let me know if you have any more questions! Happy coding! 🚀

**Comment by supriamir** [comment]
@glenn-jocher thank you for your answer.

So now the formula is (80x80 x (5+C)) for small detection head. right?

**Comment by Tom-Teamo** [comment]
@glenn-jocher I am confused about the `objectness score`. 

**Is there still a objectness score prediction in YOLOv8?**

In `DetectV8 Head`:
```
def forward(self, x):
        """Concatenates and returns predicted bounding boxes and class probabilities."""
        for i in range(self.nl):
            x[i] = torch.cat((self.cv2[i](x[i]), self.cv3[i](x[i])), 1)
        if self.training:  # Training path
            return x
````
it seems no `objectness score`.

**Comment by glenn-jocher** [comment]
Hello @Tom-Teamo,

Yes, YOLOv8 still predicts an objectness score, which indicates the confidence that a given bounding box contains an object. The code snippet you provided shows the forward pass of the DetectV8 Head, where feature maps are processed through convolutional layers. The objectness score, along with class probabilities and bounding box coordinates, are typically encoded together in the output tensor from the final convolutional layers, not explicitly shown in the snippet you posted.

The objectness score is implicitly handled within the model's architecture and is crucial for distinguishing between background and potential objects. If you need further clarification or more details on how these scores are integrated and used during training and inference, feel free to ask!

Happy coding! 😊

**Comment by guanning03** [comment]
I have a question. So if their are more than one objects in a grid, yolov8 is not able to detect them all, is my understanding right? Thanks

**Comment by pderrenger** [comment]
Thank you for your question. YOLOv8 can detect multiple objects within a single grid cell. The model predicts bounding boxes and class probabilities for each object independently, allowing it to handle multiple objects in close proximity. If you encounter issues with detecting multiple objects, ensure you are using the latest version of the package and that your training data is properly annotated. If the problem persists, please provide more details so we can assist you further.

**Comment by zener90818** [comment]
> @uname0x96 hello,
> 
> Thank you for your question, it's a good one.
> 
> In an anchor-free approach such as used in YOLOv8, we don't have predefined anchor boxes but we still do have grid cells. Predictions are made directly, rather than as an offset from a predefined anchor box. The center of an object is predicted from within the grid cell where the object resides, not as an offset from the top-left image corner. The width and height of the object are also predicted directly but are constrained to positivity to ensure valid box dimensions.
> 
> Regarding the "Reg_Max" parameter, despite being anchor-free, YOLOv8 still requires some form of regularization for the bounding box dimensions to prevent degenerate predictions. The "Reg_Max" parameter is used to define the maximum range of width and height parameters used in the prediction. It doesn't relate to anchor boxes, rather it is a form of regulation used to constrain the directly predicted box.
> 
> I hope this clears up your confusion about the anchor-free approach used in YOLOv8. Please feel free to ask if you have any further questions. Best regards!

Hi @glenn-jocher,
Can i change the regularization parameters of bbox(x,y,w,h) to circle(x,y,radius), i want to circle the eyeball to locate the pupil center. Without the bounding boxs infomation, how to dealing with the positive and negative sample distribution problem？
Thanks for your help.

**Comment by glenn-jocher** [comment]
Hi @zener90818,

YOLOv8 is primarily designed for bounding box predictions, so adapting it for circular predictions would require significant modifications to the model architecture and loss functions. For your specific use case, you might consider exploring custom models or frameworks that natively support circular or keypoint-based detections. If you have further questions or need more guidance, feel free to ask!

**Comment by leethologica** [comment]
> @supriamir hi there! 😊 Great question!
> 
> In YOLOv8, which uses an anchor-free approach, the concept of 'B' or the number of anchor boxes doesn't directly apply as it does in anchor-based models. Instead, YOLOv8 directly predicts box coordinates without predefined anchors.

As I understand it, while the inferences are anchor-free in that the output detections are defined by their center rather than an offset to pre-specified static anchors, there is still sort of a concept of anchors within the grid cells during training.

Given this diagram detect block of YOLOv8 which I have highlighted the class/box branch splits among three heads, is there a way I could determine which feature vector(s) correspond to a particular bbox output?

![Screenshot 2024-11-13 at 6 34 34 PM](https://github.com/user-attachments/assets/a3640c55-fc63-40c9-a08d-ab34d82bddf0)

That is, suppose I have extracted the feature maps from each colored head, so i have features of shape `[n, 192, 80, 80]`, `[n, 384, 40, 40]`, and `[n, 192, 20, 20]`. As I understand it, there are 8400 cells then (80\*80 + 40\*40 + 20\*20) that these features correspond to. Given an `x, y, w, h` bbox, am I correct in thinking that if I find cell center from those 8400 cells that is closest to the center of the bbox, then I can find the exact feature vector that corresponds to that detection?

For example, suppose the center of my $i$-th bbox is closest to the 1337th cell which corresponds to the 16x57 cell from the `[n, 192, 80, 80]` features, then would `[i, :, 16, 57]` be the feature vector of that prediction?

Apologies if I am way off base on my understanding of this, just trying to determine how I could find specific features that contribute most to a particular detection.
